include/mqtt/endpoint.hpp

```diff
@@ -8278,1217 +8319,3822 @@ class endpoint : public std::enable_shared_from_this<endpoint<Socket, Mutex, Loc
-                    return false;
+                    buf.remove_prefix(1);
+                    handler(
+                        size,
+                        std::move(buf),
+                        std::move(func),
+                        std::move(self)
+                    );
                 }
-                i += static_cast<std::size_t>(std::distance(b, it));
-            }
-
-            if (remaining_length_ < i + 2) {
-                if (func) func(boost::system::errc::make_error_code(boost::system::errc::message_size));
-                return false;
-            }
-            const std::uint16_t topic_name_length = make_uint16_t(payload_[i], payload_[i + 1]);
-            i += 2;
-
-            if (remaining_length_ < i + topic_name_length) {
-                if (func) func(boost::system::errc::make_error_code(boost::system::errc::message_size));
-                return false;
-            }
-            std::string topic_name(payload_.data() + i, topic_name_length);
-            if (utf8string::validate_contents(topic_name) != utf8string::validation::well_formed) {
-                if (func) func(boost::system::errc::make_error_code(boost::system::errc::bad_message));
-                return false;
-            }
-
-            i += topic_name_length;
-
-            if (remaining_length_ < i + 2) {
-                if (func) func(boost::system::errc::make_error_code(boost::system::errc::message_size));
-                return false;
-            }
-            std::uint16_t will_message_length = make_uint16_t(payload_[i], payload_[i + 1]);
-            i += 2;
+            };
 
-            if (remaining_length_ < i + will_message_length) {
-                if (func) func(boost::system::errc::make_error_code(boost::system::errc::message_size));
-                return false;
-            }
-            std::string will_message(payload_.data() + i, will_message_length);
-            i += will_message_length;
-            w = will(std::move(topic_name),
-                     std::move(will_message),
-                     connect_flags::has_will_retain(byte8),
-                     connect_flags::will_qos(byte8),
-                     std::move(will_props));
+        if (buf.empty()) {
+            socket_->async_read(
+                as::buffer(buf_.data(), 1),
+                [
+                    this,
+                    self = std::move(self),
+                    func = std::move(func),
+                    handler = std::move(handler),
+                    size,
+                    multiplier,
+                    proc = std::move(proc)
+                ]
+                (boost::system::error_code const& ec,
+                 std::size_t bytes_transferred) mutable {
+                    if (!check_error_and_transferred_length(ec, func, bytes_transferred, 1)) return;
+                    proc(
+                        std::move(func),
+                        buffer(string_view(buf_.data(), 1)),
```

Comment:
Make a comment saying the lifetime of `buf_` is handled by `self`

Reply:
I will add comment.